pkg/pool/sampler_test.go

```diff
@@ -0,0 +1,67 @@
+package pool
+
+import (
+	"math"
+	"math/rand"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestSamplerNoStability(t *testing.T) {
+	n := 100000
+	nf := float64(n / 10)
+
+	cases := [][]float64{
+		{1, 0},
+		{0.1, 0.2, 0.7},
+		{0.2, 0.2, 0.4, 0.1, 0.1, 0},
+		{0.5, 0.5},
+	}
+
+	for _, probabilities := range cases {
+		sampler := NewSampler(probabilities, rand.NewSource(time.Now().UnixNano()))
+		res := make([]float64, len(probabilities))
+		for i := 0; i < n; i++ {
+			res[sampler.Next()]++
+		}
+
+		for i, re := range res {
+			actual := math.Round(re / nf)
+			expected := math.Round(probabilities[i] * 10)
+
+			require.Equal(t, expected, actual, "probabilities: %v, result: %v", probabilities, res)
+		}
+	}
+}
+
+func TestSamplerStability(t *testing.T) {
+	cases := []struct {
+		probabilities []float64
+		expected      []int
+	}{
+		{
+			probabilities: []float64{1, 0},
+			expected:      []int{100000, 0},
+		},
+		{
+			probabilities: []float64{0.1, 0.2, 0.7},
+			expected:      []int{10138, 19813, 70049},
+		},
+		{
+			probabilities: []float64{0.2, 0.2, 0.4, 0.1, 0.1, 0},
+			expected:      []int{19824, 20169, 39900, 10243, 9864, 0},
+		},
+	}
+
+	for _, tc := range cases {
+		sampler := NewSampler(tc.probabilities, rand.NewSource(0))
+		res := make([]int, len(tc.probabilities))
+		for i := 0; i < 100000; i++ {
```

Comment:
This number is related to expected numbers, so we better declare it somewhere above as `const`.